core/src/main/java/feast/core/service/JobCoordinatorService.java

```diff
@@ -177,16 +158,145 @@ private void updateFeatureSetStatuses(List<JobUpdateTask> jobUpdateTasks) {
+          isSafeToStopJobs = false;
+        }
+      }
+
+      jobUpdateTasks.add(
+          new JobUpdateTask(
+              getFeatureSetsForStore(store),
+              source,
+              store,
+              sourceToStoreJob,
+              jobManager,
+              updateTimeout,
+              JobStatus.RUNNING));
+    }
+    // Stop extra jobs that are not required to mantain ingestion when safe
+    if (isSafeToStopJobs) {
+      getExtraJobs(requiredJobsIds)
+          .forEach(
+              extraJob -> {
+                jobUpdateTasks.add(
+                    new JobUpdateTask(
+                        extraJob.getFeatureSets(),
+                        extraJob.getSource(),
+                        extraJob.getStore(),
+                        Optional.of(extraJob),
+                        jobManager,
+                        updateTimeout,
+                        JobStatus.ABORTED));
+              });
+    }
+
+    return jobUpdateTasks;
+  }
+
+  /** Get the non terminated ingestion job ingesting between source and source if exists. */
   @Transactional
   public Optional<Job> getJob(Source source, Store store) {
-    List<Job> jobs =
-        jobRepository.findBySourceIdAndStoreNameOrderByLastUpdatedDesc(
-            source.getId(), store.getName());
-    jobs = jobs.stream().filter(job -> !job.hasTerminated()).collect(Collectors.toList());
-    if (jobs.isEmpty()) {
-      return Optional.empty();
+    return jobRepository.findFirstBySourceIdAndStoreNameAndStatusNotInOrderByLastUpdatedDesc(
+        source.getId(), store.getName(), JobStatus.getTerminalStates());
+  }
+
+  /** Get non terminated extra ingestion jobs that have ids not in keepJobIds */
+  @Transactional
+  private List<Job> getExtraJobs(List<String> keepJobIds) {
+    return jobRepository.findByIdNotInAndStatusNotIn(keepJobIds, JobStatus.getTerminalStates());
+  }
+
+  /**
+   * Generate a source to store mapping. The mapping maps the source/store pairs in which ingestion
+   * jobs would have to be maintained for ingestion to work correctly.
+   *
+   * @return a map of source to store.
+   */
+  private Map<Source, Store> getSourceToStoreMapping() {
+    ListStoresResponse listStoresResponse = specService.listStores(Filter.newBuilder().build());
+    List<Store> stores =
+        listStoresResponse.getStoreList().stream()
+            .map(storeSpec -> Store.fromProto(storeSpec))
+            .collect(Collectors.toList());
+
+    // compile a set of sources via subscribed FeatureSets of Stores.
+    Set<Source> sources =
+        stores.stream()
+            .flatMap(
+                store ->
+                    getFeatureSetsForStore(store).stream()
+                        .map(featureSet -> featureSet.getSource()))
+            .distinct()
+            .map(source -> getSurogateSource(source))
+            .collect(Collectors.toSet());
+
+    Map<Source, Store> sourceStoreMap = new HashMap<>();
+    for (Source source : sources) {
```

Comment:
On this point we may generate a pair that actually doesn't have a connection. Source may be connected only to some stores, not all of them, for example. But it definitely not all stores X all sources.
Why not continue source compilation (previous line) with pair generation?
```
.flatMap(
  store ->
    getFeatureSetsForStore(store).stream()
         .map(featureSet -> Pair.of(featureSet.getSource(), store)))
.distinct()
.collect()
```
I don't quit get why we need surrogate source here

Reply:
Correction, since it's a map we overwriting stores for the same source key and thus end up with one store per source.
Should be list of pairs

Reply:
Corrected. Thanks for pointing this out.

Reply:
My first comment is still valid. It shouldn't be all sources X all stores. You need to create pairs that are actually connected.
I guess to see that's the case would be great to add test that creates
Store1 -> subscripted only to -> Source1
Store2 -> subscripted only to -> Source2
Only two pairs should be generated. Whereas current implementation will generate 4

Reply:
Corrected.